crates/node-api/src/payload_attributes.rs

```diff
@@ -0,0 +1,186 @@
+use crate::{validate_version_specific_fields, AttributesValidationError, EngineApiMessageVersion};
+use reth_primitives::{
+    revm::config::revm_spec_by_timestamp_after_merge,
+    revm_primitives::{BlobExcessGasAndPrice, BlockEnv, CfgEnv, SpecId},
+    Address, ChainSpec, Header, B256, U256,
+};
+use reth_rpc_types::engine::{
+    OptimismPayloadAttributes, PayloadAttributes as EthPayloadAttributes, PayloadId, Withdrawal,
+};
+
+/// This can be implemented by types that describe a currently running payload job.
+pub trait PayloadBuilderAttributes {
+    /// The payload attributes that can be used to construct this type. Used as the argument in
+    /// [PayloadBuilderAttributes::try_new].
+    type RpcPayloadAttributes;
+    /// The error type used in [PayloadBuilderAttributes::try_new].
+    type Error: std::error::Error;
+
+    /// Creates a new payload builder for the given parent block and the attributes.
+    ///
+    /// Derives the unique [PayloadId] for the given parent and attributes
+    fn try_new(
+        parent: B256,
+        rpc_payload_attributes: Self::RpcPayloadAttributes,
+    ) -> Result<Self, Self::Error>
+    where
+        Self: Sized;
+
+    /// Returns the [PayloadId] for the running payload job.
+    fn payload_id(&self) -> PayloadId;
+
+    /// Returns the parent block hash for the running payload job.
+    fn parent(&self) -> B256;
+
+    /// Returns the timestmap for the running payload job.
+    fn timestamp(&self) -> u64;
+
+    /// Returns the parent beacon block root for the running payload job, if it exists.
+    fn parent_beacon_block_root(&self) -> Option<B256>;
+
+    /// Returns the suggested fee recipient for the running payload job.
+    fn suggested_fee_recipient(&self) -> Address;
+
+    /// Returns the prevrandao field for the running payload job.
+    fn prev_randao(&self) -> B256;
+
+    /// Returns the withdrawals for the running payload job.
+    fn withdrawals(&self) -> &Vec<reth_primitives::Withdrawal>;
+
+    /// Returns the configured [CfgEnv] and [BlockEnv] for the targeted payload (that has the
+    /// `parent` as its parent).
+    ///
+    /// The `chain_spec` is used to determine the correct chain id and hardfork for the payload
+    /// based on its timestamp.
+    ///
+    /// Block related settings are derived from the `parent` block and the configured attributes.
+    ///
+    /// NOTE: This is only intended for beacon consensus (after merge).
+    fn cfg_and_block_env(&self, chain_spec: &ChainSpec, parent: &Header) -> (CfgEnv, BlockEnv) {
+        // TODO: should be different once revm has configurable cfgenv
+        // configure evm env based on parent block
+        let mut cfg = CfgEnv::default();
+        cfg.chain_id = chain_spec.chain().id();
+
+        #[cfg(feature = "optimism")]
+        {
+            cfg.optimism = chain_spec.is_optimism();
+        }
```

Comment:
this is fine for now, because eventually we need to change this anyway once we have evm builder